### PR TITLE
chore(refactor): separate out instance name into own class

### PR DESCRIPTION
**Pull request by shubha-rajan**

### CODE REVIEW

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 1]
```diff
@@ -138,10 +138,9 @@ class CloudSqlInstance {
             "[%s] Cloud SQL connection name is invalid, expected string in the form of"
                 + " \"<PROJECT_ID>:<REGION_ID>:<INSTANCE_ID>\".",
             connectionName));
-    this.connectionName = connectionName;
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:141]: Let's move the matcher into the CloudSqlInstanceName constructor.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java [hunk 2]
```diff
@@ -532,33 +532,35 @@ private SslData createSslData(
   private Metadata fetchMetadata() {
     try {
       ConnectSettings instanceMetadata =
-          apiClient.connect().get(projectId, instanceId).execute();
+          apiClient.connect().get(
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstance.java:535]: Let's match the Google Style guide and only break lines if they're over 100 characters.

**Review by enocom — CHANGES_REQUESTED** [review]

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CloudSqlInstanceName {
+
+  // Unique identifier for each Cloud SQL instance in the format "PROJECT:REGION:INSTANCE"
+  // Some legacy project ids are domain-scoped (e.g. "example.com:PROJECT:REGION:INSTANCE")
+  private static final Pattern CONNECTION_NAME =
+      Pattern.compile("([^:]+(:[^:]+)?):([^:]+):([^:]+)");
+  private String projectId;
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java:30]: Project, region, and instance can all be final.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class CloudSqlInstanceName {
+
+  // Unique identifier for each Cloud SQL instance in the format "PROJECT:REGION:INSTANCE"
+  // Some legacy project ids are domain-scoped (e.g. "example.com:PROJECT:REGION:INSTANCE")
+  private static final Pattern CONNECTION_NAME =
+      Pattern.compile("([^:]+(:[^:]+)?):([^:]+):([^:]+)");
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java:29]: Let's pull this up and use 100 characters across the board.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.sql.core;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+import java.util.regex.Matcher;
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java:21]: There's a [google-java-format](https://plugins.jetbrains.com/plugin/8527-google-java-format) plugin for IntelliJ that will do this automatically for you, but these imports should appear above the checkArgument import.

**Review by enocom — CHANGES_REQUESTED** [review]
A few minor things and then this is ready to go.

**Comment by enocom** [comment]
Let's also plan to fix the various warnings here as we go: https://github.com/GoogleCloudPlatform/cloud-sql-jdbc-socket-factory/actions/runs/4318203517/jobs/7536211361.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java (hunk 4 above)
Comment (shubha-rajan) [reply core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java:29]: This line is 102 lines if we make it one line 🙃

Comment (enocom) [reply core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java]: oops.

File: core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+import java.util.regex.Pattern;
+
+/**
+ * This class parses the different parts of a Cloud SQL Connection Name to allow users to easily
+ * fetch the projectId, regionId, and instanceId.
+ */
+public class CloudSqlInstanceName {
+
+  // Unique identifier for each Cloud SQL instance in the format "PROJECT:REGION:INSTANCE"
+  // Some legacy project ids are domain-scoped (e.g. "example.com:PROJECT:REGION:INSTANCE")
```
Comment (enocom) [inline core/src/main/java/com/google/cloud/sql/core/CloudSqlInstanceName.java:31]: Shall we write a couple of unit tests to codify this behavior?

**Review by enocom — APPROVED** [review]

**Pull request merged**